t/ui/23-audit-log.t

```diff
@@ -58,55 +52,80 @@ is($driver->find_element('#user-action a')->get_text(), 'Logged in as Demo', "lo
 
 $driver->find_element('#user-action a')->click();
 $driver->find_element_by_link_text('Audit log')->click();
-wait_for_data_table;
+my $table = wait_for_element(selector => '#audit_log_table');
 like($driver->get_title(), qr/Audit log/, 'on audit log');
-my $table = $driver->find_element_by_id('audit_log_table');
-ok($table, 'audit table found');
 
 # search for name, event, date and combination
 my $search = $driver->find_element('#audit_log_table_filter input.form-control');
 ok($search, 'search box found');
-
-my @entries = $driver->find_child_elements($table, 'tbody/tr', 'xpath');
-is(scalar @entries, 3, 'three elements without filter');
+wait_for_element(
+    selector    => '#audit_log_table tbody tr:nth-child(3):last-child',
+    description => 'three elements without filter'
+);
 
 $search->send_keys('QA restart');
-wait_for_data_table;
-@entries = $driver->find_child_elements($table, 'tbody/tr', 'xpath');
-is(scalar @entries, 1, 'one element when filtered for event data');
-like($entries[0]->get_text(), qr/openQA restarted/, 'correct element displayed');
+my $element = wait_for_element(selector => '#audit_log_table tbody tr:only-child', description => 'one element');
+like($element->get_text(), qr/openQA restarted/, 'correct element displayed');
 $search->clear;
 
 $search->send_keys('user:system');
-wait_for_data_table;
-@entries = $driver->find_child_elements($table, 'tbody/tr', 'xpath');
-is(scalar @entries, 1, 'one element when filtered by user');
+wait_for_element(selector => '#audit_log_table tbody tr:only-child', description => 'one element');
 $search->clear;
 
 $search->send_keys('event:user_login');
-wait_for_data_table;
-@entries = $driver->find_child_elements($table, 'tbody/tr', 'xpath');
-is(scalar @entries, 2, 'two elements when filtered by event');
+wait_for_element(
+    selector    => '#audit_log_table tbody tr:nth-child(2):last-child',
+    description => 'two elements for event'
+);
 $search->clear;
 
 $search->send_keys('newer:today');
-wait_for_data_table;
-@entries = $driver->find_child_elements($table, 'tbody/tr', 'xpath');
-is(scalar @entries, 3, 'three elements when filtered by today time');
+wait_for_element(
+    selector    => '#audit_log_table tbody tr:nth-child(3):last-child',
+    description => 'three elements for today'
+);
 $search->clear;
 
 $search->send_keys('older:today');
-wait_for_data_table;
-@entries = $driver->find_child_elements($table, 'tbody/tr/td', 'xpath');
-is(scalar @entries,                     1,                  'one element when filtered by yesterday time');
-is($entries[0]->get_attribute('class'), 'dataTables_empty', 'but datatables are empty');
+wait_for_element(
```

Comment:
I assume the regular `wait_for_ajax` didn't work here?

Reply:
It "worked" before by using a hard-coded `sleep`. Otherwise it returns right away and the element isn't there.
Waiting for the actual element(s) doesn't have that problem.

Reply:
It might be that DataTables gives the control back to the event loop early breaking the synchronization. I guess then we need to use `wait_for_element`, indeed.